File: jarvis-personal/native/android/core/data/src/main/kotlin/com/dincr/data/DincrService.kt
package com.dincr.data

import java.net.URLEncoder

/** Backend operations the native app uses; 1:1 with the Capacitor app's routes (parity IDs). */
interface DincrService {
    /** A5, A14 */ suspend fun me(): Profile
    /** A9 */ suspend fun completeProfileSetup(setup: ProfileSetup): Profile
    /** C1 */ suspend fun freeDashboard(): FreeDashboard
    /** D1, D5 */ suspend fun movements(): List<Movement>
    /** D3. [idempotencyKey] identifies one user submission: sending it again cannot create a second row. */
    suspend fun create(kind: MovementKind, entry: EntryCreate, idempotencyKey: String)
    /** D4, D5 */ suspend fun update(movementId: String, update: MovementUpdate)
    /** D6 */ suspend fun delete(movementId: String)
}

@kotlinx.serialization.Serializable
internal data class Acknowledgement(val status: String? = null)

class LiveDincrService(private val client: ApiClient) : DincrService {
    override suspend fun me(): Profile = client.get("/auth/me")
    override suspend fun completeProfileSetup(setup: ProfileSetup): Profile =
        client.send<ProfileSetup, ProfileEnvelope>("POST", "/auth/profile-setup", setup).profile
    override suspend fun freeDashboard(): FreeDashboard = client.get("/user-product/free/dashboard")
    override suspend fun movements(): List<Movement> = client.get("/user-product/free/movements")
    override suspend fun create(kind: MovementKind, entry: EntryCreate, idempotencyKey: String) {
        val path = if (kind == MovementKind.INCOME) "/user-product/finance/income" else "/user-product/finance/expenses"
        client.send<EntryCreate, Acknowledgement>("POST", path, entry, idempotencyKey)
    }
    override suspend fun update(movementId: String, update: MovementUpdate) {
        client.send<MovementUpdate, Acknowledgement>("PUT", "/user-product/free/movements/${encode(movementId)}", update)
    }
    override suspend fun delete(movementId: String) {
        client.send<Acknowledgement>("DELETE", "/user-product/free/movements/${encode(movementId)}")
    }

    companion object {
        /** Movement ids look like `expense:42`; keep the colon, escape anything path-breaking. */
        fun encode(id: String): String = URLEncoder.encode(id, "UTF-8").replace("%3A", ":").replace("+", "%20")
    }
}

File: jarvis-personal/native/ios/DincrKit/Sources/DincrCore/DincrService.swift
import Foundation

/// The backend operations the native app uses. Each maps 1:1 to a FastAPI route consumed by the
/// Capacitor app (parity IDs in docs/native/PARITY_MATRIX.md).
public protocol DincrService: Sendable {
    /// A5, A14 — `GET /auth/me`
    func me() async throws -> Profile
    /// A9 — `POST /auth/profile-setup`
    func completeProfileSetup(_ setup: ProfileSetup) async throws -> Profile
    /// C1 — `GET /user-product/free/dashboard`
    func freeDashboard() async throws -> FreeDashboard
    /// D1, D5 — `GET /user-product/free/movements`
    func movements() async throws -> [Movement]
    /// D3 — `POST /user-product/finance/income` | `/expenses`
    func create(_ kind: Movement.Kind, _ entry: EntryCreate) async throws
    /// D4, D5 — `PUT /user-product/free/movements/{id}`
    func update(movementID: String, _ update: MovementUpdate) async throws
    /// D6 — `DELETE /user-product/free/movements/{id}`
    func delete(movementID: String) async throws
}

public struct LiveDincrService: DincrService {
    let client: APIClient

    public init(client: APIClient) { self.client = client }

    public func me() async throws -> Profile { try await client.get("/auth/me") }

    public func completeProfileSetup(_ setup: ProfileSetup) async throws -> Profile {
        let envelope: ProfileEnvelope = try await client.send("POST", "/auth/profile-setup", body: setup)
        return envelope.profile
    }

    public func freeDashboard() async throws -> FreeDashboard { try await client.get("/user-product/free/dashboard") }

    public func movements() async throws -> [Movement] { try await client.get("/user-product/free/movements") }

    public func create(_ kind: Movement.Kind, _ entry: EntryCreate) async throws {
        let path = kind == .income ? "/user-product/finance/income" : "/user-product/finance/expenses"
        let _: Acknowledgement = try await client.send("POST", path, body: entry)
    }

    public func update(movementID: String, _ update: MovementUpdate) async throws {
        let _: Acknowledgement = try await client.send("PUT", "/user-product/free/movements/\(Self.encode(movementID))", body: update)
    }

    public func delete(movementID: String) async throws {
        let _: Acknowledgement = try await client.send("DELETE", "/user-product/free/movements/\(Self.encode(movementID))")
    }

    /// Movement ids look like `expense:42`; the colon must survive as a path segment.
    static func encode(_ id: String) -> String {
        id.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed.subtracting(CharacterSet(charactersIn: "/"))) ?? id
    }
}
